controllers/hyperparametertuningjob/hyperparametertuningjob_controller_test.go

```diff
@@ -18,1680 +18,538 @@ package hyperparametertuningjob
-			TrainingJobName: &expectedTrainingJobName,
-		}
-
-		// Setup mock responses.
-		sageMakerClient := builder.
-			AddDescribeHyperParameterTuningJobResponse(description).
-			Build()
-
-		// Create controller with mocks and its request
-		controller := createHpoReconciler(k8sClient, CreateMockSageMakerClientProvider(sageMakerClient), createMockHpoTrainingJobSpawnerProvider(noopHpoTrainingJobSpawner{}), CreateMockAwsConfigLoader(), "1s")
-		request := CreateReconciliationRequest(job.ObjectMeta.Name, job.ObjectMeta.Namespace)
+	Context("HyperParameterTuningJob does not exist", func() {
 
-		// Run test
-		_, err := controller.Reconcile(request)
-
-		Expect(err).ToNot(HaveOccurred())
+		BeforeEach(func() {
+			mockSageMakerClientBuilder.
+				AddDescribeHyperParameterTuningJobErrorResponse(clientwrapper.DescribeHyperParameterTuningJob404Code, clientwrapper.DescribeHyperParameterTuningJob404MessagePrefix, 400, "request id")
+		})
 
-		// Get job so we can verify it has the correct status.
-		err = k8sClient.Get(context.Background(), types.NamespacedName{
-			Namespace: job.ObjectMeta.Namespace,
-			Name:      job.ObjectMeta.Name,
-		}, job)
-		Expect(err).ToNot(HaveOccurred())
+		Context("HasDeletionTimestamp", func() {
 
-		Expect(job.Status.BestTrainingJob.TrainingJobName).ToNot(BeNil())
-		Expect(*job.Status.BestTrainingJob.TrainingJobName).To(Equal(expectedTrainingJobName))
-	})
+			BeforeEach(func() {
+				shouldHaveDeletionTimestamp = true
+				shouldHaveFinalizer = true
+			})
 
-	It("should update the status TrainingJobStatusCounters", func() {
-		// Set status
-		var completed int64 = 1
-		var inProgress int64 = 2
-		var nonRetryableError int64 = 3
-		var retryableError int64 = 4
-		var stopped int64 = 5
-
-		// Set mock error values.
-		description.TrainingJobStatusCounters = &sagemaker.TrainingJobStatusCounters{
-			Completed:         &completed,
-			InProgress:        &inProgress,
-			NonRetryableError: &nonRetryableError,
-			RetryableError:    &retryableError,
-			Stopped:           &stopped,
-		}
+			It("Removes finalizer and deletes HyperParameterTuningJob", func() {
+				ExpectHyperParameterTuningJobToBeDeleted(tuningJob)
+			})
 
-		// Setup mock responses.
-		sageMakerClient := builder.
-			AddDescribeHyperParameterTuningJobResponse(description).
-			Build()
+			It("Requeues after interval", func() {
+				ExpectRequeueAfterInterval(reconcileResult, reconcileError, pollDuration)
+			})
+		})
 
-		// Create controller with mocks and its request
-		controller := createHpoReconciler(k8sClient, CreateMockSageMakerClientProvider(sageMakerClient), createMockHpoTrainingJobSpawnerProvider(noopHpoTrainingJobSpawner{}), CreateMockAwsConfigLoader(), "1s")
-		request := CreateReconciliationRequest(job.ObjectMeta.Name, job.ObjectMeta.Namespace)
+		Context("!HasDeletionTimestamp", func() {
+			BeforeEach(func() {
+				mockSageMakerClientBuilder.
+					AddCreateHyperParameterTuningJobResponse(sagemaker.CreateHyperParameterTuningJobOutput{}).
```

Comment:
We should have tests that assert any failure to create is propagated to the user via the `Status.Additional` field. This is particularly important because any s3 region error, iam error, capacity error, et al will surface to customer this way.

Reply:
This seems like a unit test addition for every controller. Perhaps lets break this into another refactor task?

Reply:
++, I'll add this comment to https://issues.amazon.com/issues/P31113513